// react-bnb-gallery/dist/react-bnb-gallery.es.js
import forbidExtraProps from 'airbnb-prop-types/src/forbidExtraProps';
import 'airbnb-prop-types/src/mutuallyExclusiveProps';
import nonNegativeInteger from 'airbnb-prop-types/src/nonNegativeInteger';
import PropTypes from 'prop-types';
import React, { PureComponent, Component } from 'react';
import classnames from 'classnames';
import omit from 'lodash/omit';
import FocusTrap from 'focus-trap-react';
import { Portal } from 'react-portal';

function _extends() {
  _extends = Object.assign || function (target) {
    for (var i = 1; i < arguments.length; i++) {
      var source = arguments[i];

      for (var key in source) {
        if (Object.prototype.hasOwnProperty.call(source, key)) {
          target[key] = source[key];
        }
      }
    }

    return target;
  };

  return _extends.apply(this, arguments);
}

function _classCallCheck(instance, Constructor) {
  if (!(instance instanceof Constructor)) {
    throw new TypeError("Cannot call a class as a function");
  }
}

function _defineProperties(target, props) {
  for (var i = 0; i < props.length; i++) {
    var descriptor = props[i];
    descriptor.enumerable = descriptor.enumerable || false;
    descriptor.configurable = true;
    if ("value" in descriptor) descriptor.writable = true;
    Object.defineProperty(target, descriptor.key, descriptor);
  }
}

function _createClass(Constructor, protoProps, staticProps) {
  if (protoProps) _defineProperties(Constructor.prototype, protoProps);
  if (staticProps) _defineProperties(Constructor, staticProps);
  return Constructor;
}

function _typeof2(obj) { if (typeof Symbol === "function" && typeof Symbol.iterator === "symbol") { _typeof2 = function _typeof2(obj) { return typeof obj; }; } else { _typeof2 = function _typeof2(obj) { return obj && typeof Symbol === "function" && obj.constructor === Symbol && obj !== Symbol.prototype ? "symbol" : typeof obj; }; } return _typeof2(obj); }

function _typeof(obj) {
  if (typeof Symbol === "function" && _typeof2(Symbol.iterator) === "symbol") {
    _typeof = function _typeof(obj) {
      return _typeof2(obj);
    };
  } else {
    _typeof = function _typeof(obj) {
      return obj && typeof Symbol === "function" && obj.constructor === Symbol && obj !== Symbol.prototype ? "symbol" : _typeof2(obj);
    };
  }

  return _typeof(obj);
}

function _assertThisInitialized(self) {
  if (self === void 0) {
    throw new ReferenceError("this hasn't been initialised - super() hasn't been called");
  }

  return self;
}

function _possibleConstructorReturn(self, call) {
  if (call && (_typeof(call) === "object" || typeof call === "function")) {
    return call;
  }

  return _assertThisInitialized(self);
}

function _getPrototypeOf(o) {
  _getPrototypeOf = Object.setPrototypeOf ? Object.getPrototypeOf : function _getPrototypeOf(o) {
    return o.__proto__ || Object.getPrototypeOf(o);
  };
  return _getPrototypeOf(o);
}

function _setPrototypeOf(o, p) {
  _setPrototypeOf = Object.setPrototypeOf || function _setPrototypeOf(o, p) {
    o.__proto__ = p;
    return o;
  };

  return _setPrototypeOf(o, p);
}

function _inherits(subClass, superClass) {
  if (typeof superClass !== "function" && superClass !== null) {
    throw new TypeError("Super expression must either be null or a function");
  }

  subClass.prototype = Object.create(superClass && superClass.prototype, {
    constructor: {
      value: subClass,
      writable: true,
      configurable: true
    }
  });
  if (superClass) _setPrototypeOf(subClass, superClass);
}

function _defineProperty(obj, key, value) {
  if (key in obj) {
    Object.defineProperty(obj, key, {
      value: value,
      enumerable: true,
      configurable: true,
      writable: true
    });
  } else {
    obj[key] = value;
  }

  return obj;
}

function _objectSpread(target) {
  for (var i = 1; i < arguments.length; i++) {
    var source = arguments[i] != null ? arguments[i] : {};
    var ownKeys = Object.keys(source);

    if (typeof Object.getOwnPropertySymbols === 'function') {
      ownKeys = ownKeys.concat(Object.getOwnPropertySymbols(source).filter(function (sym) {
        return Object.getOwnPropertyDescriptor(source, sym).enumerable;
      }));
    }

    ownKeys.forEach(function (key) {
      _defineProperty(target, key, source[key]);
    });
  }

  return target;
}

var DIRECTION_PREV = 'prev';
var DIRECTION_NEXT = 'next';
var FORWARDS = 'forwards';
var BACKWARDS = 'backwards';
var ARROW_LEFT_KEYCODE = 37;
var ARROW_RIGHT_KEYCODE = 39;
var ESC_KEYCODE = 27;
var THUMBNAIL_WIDTH = 100;
var THUMBNAIL_HEIGHT = 67;
var THUMBNAIL_OFFSET = 10;
var DEFAULT_OPACITY = 1;
var DEFAULT_COLOR = 'rgba(0,0,0,1)';
var NORMAL_COLOR = 'rgba(255,255,255,1)';
var INVERSE_COLOR = 'rgba(1,1,1,1)';
var DEFAULT_Z_INDEX = 2000;
var defaultPhotoProps = {
  photo: undefined,
  number: undefined,
  caption: undefined,
  subcaption: undefined,
  thumbnail: undefined
};

var SlideDirectionShape = PropTypes.oneOf([FORWARDS, BACKWARDS]);

var PhotoShape = PropTypes.shape({
  photo: PropTypes.string.isRequired,
  number: PropTypes.number,
  caption: PropTypes.string,
  subcaption: PropTypes.string,
  thumbnail: PropTypes.string
});

var PhotosShape = PropTypes.oneOfType([PropTypes.arrayOf(PropTypes.string.isRequired), PropTypes.arrayOf(PhotoShape)]);

var noPhotosProvided = 'No photos to show';
var showPhotoList = 'Show photo list';
var hidePhotoList = 'Hide photo list';
var defaultPhrases = {
  noPhotosProvided: noPhotosProvided,
  showPhotoList: showPhotoList,
  hidePhotoList: hidePhotoList
};

function defaultPhrases$1() {
  return Object.keys(defaultPhrases$1).reduce(function (phrases, key) {
    return _objectSpread({}, phrases, _defineProperty({}, key, PropTypes.oneOfType([PropTypes.string, PropTypes.func, PropTypes.node])));
  }, {});
}

var galleryPropTypes = forbidExtraProps({
  activePhotoIndex: nonNegativeInteger,
  activePhotoPressed: PropTypes.func,
  direction: SlideDirectionShape,
  nextButtonPressed: PropTypes.func,
  prevButtonPressed: PropTypes.func,
  showThumbnails: PropTypes.bool,
  photos: PhotosShape,
  preloadSize: nonNegativeInteger,
  wrap: PropTypes.bool,
  phrases: PropTypes.shape(defaultPhrases$1(defaultPhrases)),
  light: PropTypes.bool
});

var noop = (function () {});

var galleryDefaultProps = {
  activePhotoIndex: 0,
  activePhotoPressed: noop,
  direction: FORWARDS,
  nextButtonPressed: noop,
  prevButtonPressed: noop,
  showThumbnails: true,
  photos: [],
  preloadSize: 5,
  wrap: false,
  phrases: defaultPhrases,
  light: false
};

var imagePropTypes = forbidExtraProps({
  onLoad: PropTypes.func,
  onError: PropTypes.func
});

var imageDefaultProps = {
  onLoad: noop,
  onError: noop
};

function _objectWithoutPropertiesLoose(source, excluded) {
  if (source == null) return {};
  var target = {};
  var sourceKeys = Object.keys(source);
  var key, i;

  for (i = 0; i < sourceKeys.length; i++) {
    key = sourceKeys[i];
    if (excluded.indexOf(key) >= 0) continue;
    target[key] = source[key];
  }

  return target;
}

function _objectWithoutProperties(source, excluded) {
  if (source == null) return {};
  var target = _objectWithoutPropertiesLoose(source, excluded);
  var key, i;

  if (Object.getOwnPropertySymbols) {
    var sourceSymbolKeys = Object.getOwnPropertySymbols(source);

    for (i = 0; i < sourceSymbolKeys.length; i++) {
      key = sourceSymbolKeys[i];
      if (excluded.indexOf(key) >= 0) continue;
      if (!Object.prototype.propertyIsEnumerable.call(source, key)) continue;
      target[key] = source[key];
    }
  }

  return target;
}

var _jsxFileName = "O:\\www\\www\\react\\react-bnb-gallery\\src\\components\\LoadingSpinner.js";
var propTypes = forbidExtraProps({
  show: PropTypes.bool
});
var defaultProps = {
  show: true
};

var LoadingSpinner = function LoadingSpinner(_ref) {
  var show = _ref.show;
  return show && React.createElement("div", {
    className: "loading-spinner",
    __source: {
      fileName: _jsxFileName,
      lineNumber: 15
    },
    __self: this
  });
};

LoadingSpinner.propTypes = propTypes;
LoadingSpinner.defaultProps = defaultProps;

var _jsxFileName$1 = "O:\\www\\www\\react\\react-bnb-gallery\\src\\components\\Image.js";
var propTypes$1 = forbidExtraProps(_objectSpread({}, imagePropTypes, {
  alt: PropTypes.string.isRequired,
  src: PropTypes.string.isRequired,
  style: PropTypes.object,
  className: PropTypes.oneOfType([PropTypes.string, PropTypes.array])
}));

var defaultProps$1 = _objectSpread({}, imageDefaultProps, {
  style: null,
  className: null,
  alt: ''
});

var Image =
/*#__PURE__*/
function (_PureComponent) {
  _inherits(Image, _PureComponent);

  function Image(props) {
    var _this;

    _classCallCheck(this, Image);

    _this = _possibleConstructorReturn(this, _getPrototypeOf(Image).call(this, props));
    _this.state = {
      loading: true,
      withError: false
    };
    _this.onLoad = _this.onLoad.bind(_assertThisInitialized(_assertThisInitialized(_this)));
    _this.onError = _this.onError.bind(_assertThisInitialized(_assertThisInitialized(_this)));
    return _this;
  }

  _createClass(Image, [{
    key: "onLoad",
    value: function onLoad() {
      var _this2 = this;

      var onLoad = this.props.onLoad; // wait a litle to show the final picture

      setTimeout(function () {
        onLoad();

        _this2.setState({
          loading: false,
          withError: false
        });
      }, 100);
    }
  }, {
    key: "onError",
    value: function onError() {
      var onError = this.props.onError;
      onError();
      this.setState({
        loading: false,
        withError: true
      });
    }
  }, {
    key: "renderImage",
    value: function renderImage() {
      var _this$props = this.props,
          alt = _this$props.alt,
          src = _this$props.src,
          style = _this$props.style,
          className = _this$props.className;
      var _this$state = this.state,
          loading = _this$state.loading,
          withError = _this$state.withError;
      var classNames = [className, 'media-image'];
      var components = []; // the loading spinner
      // TODO: make this 'LoadingSpinner' component customizable

      if (loading) {
        components.push(React.createElement(LoadingSpinner, {
          key: ".pictureLoadingSpinner",
          __source: {
            fileName: _jsxFileName$1,
            lineNumber: 88
          },
          __self: this
        }));
      } // if no loading, then return the
      // picture only if no error ocurred


      if (!withError) {
        components.push(React.createElement("img", {
          alt: alt,
          key: ".pictureComponent",
          className: classnames(classNames),
          onLoad: this.onLoad,
          onError: this.onError,
          src: src,
          style: style,
          __source: {
            fileName: _jsxFileName$1,
            lineNumber: 96
          },
          __self: this
        }));
      } // TODO: show a custom message indicating the
      // error ocurred while loading the picture


      return components;
    }
  }, {
    key: "render",
    value: function render() {
      var loading = this.state.loading;
      var wrapperClassNames = ['picture', loading && 'loading']; // render the picture element

      var picture = this.renderImage();
      return React.createElement("div", {
        className: classnames(wrapperClassNames),
        __source: {
          fileName: _jsxFileName$1,
          lineNumber: 128
        },
        __self: this
      }, picture);
    }
  }]);

  return Image;
}(PureComponent);

Image.propTypes = propTypes$1;
Image.defaultProps = defaultProps$1;

var _jsxFileName$2 = "O:\\www\\www\\react\\react-bnb-gallery\\src\\components\\GalleryPhoto.js";
var propTypes$2 = forbidExtraProps(_objectSpread({}, imagePropTypes, {
  photo: PhotoShape,
  onPress: PropTypes.func
}));

var defaultProps$2 = _objectSpread({}, imageDefaultProps, {
  photo: null,
  onPress: noop
});

var GalleryPhoto =
/*#__PURE__*/
function (_PureComponent) {
  _inherits(GalleryPhoto, _PureComponent);

  function GalleryPhoto(props) {
    var _this;

    _classCallCheck(this, GalleryPhoto);

    _this = _possibleConstructorReturn(this, _getPrototypeOf(GalleryPhoto).call(this, props));
    _this.onPress = _this.onPress.bind(_assertThisInitialized(_assertThisInitialized(_this)));
    return _this;
  }

  _createClass(GalleryPhoto, [{
    key: "onPress",
    value: function onPress() {
      var onPress = this.props.onPress;
      onPress();
    }
  }, {
    key: "renderPhoto",
    value: function renderPhoto() {
      var _this$props = this.props,
          photo = _this$props.photo,
          rest = _objectWithoutProperties(_this$props, ["photo"]);

      if (!photo) {
        return null;
      }

      var imageProps = omit(rest, ['onPress']);
      return React.createElement("button", {
        type: "button",
        onClick: this.onPress,
        className: "photo-button",
        __source: {
          fileName: _jsxFileName$2,
          lineNumber: 55
        },
        __self: this
      }, React.createElement(Image, _extends({
        alt: photo.caption || '',
        className: "photo",
        src: photo.photo
      }, imageProps, {
        __source: {
          fileName: _jsxFileName$2,
          lineNumber: 60
        },
        __self: this
      })));
    }
  }, {
    key: "render",
    value: function render() {
      var className = classnames('gallery-media-photo', 'gallery-media-photo--block', 'gallery-media-cover');
      var photoRendered = this.renderPhoto();
      return React.createElement("ul", {
        className: "gallery-images--ul",
        __source: {
          fileName: _jsxFileName$2,
          lineNumber: 80
        },
        __self: this
      }, React.createElement("li", {
        className: className,
        __source: {
          fileName: _jsxFileName$2,
          lineNumber: 81
        },
        __self: this
      }, photoRendered));
    }
  }]);

  return GalleryPhoto;
}(PureComponent);

GalleryPhoto.propTypes = propTypes$2;
GalleryPhoto.defaultProps = defaultProps$2;

var _jsxFileName$3 = "O:\\www\\www\\react\\react-bnb-gallery\\src\\components\\GalleryThumbnail.js";
var thumbnailStyle = {
  width: THUMBNAIL_WIDTH,
  height: THUMBNAIL_HEIGHT
};
var propTypes$3 = forbidExtraProps({
  active: PropTypes.bool,
  photo: PhotoShape,
  onPress: PropTypes.func,
  number: nonNegativeInteger
});
var defaultProps$3 = {
  active: false,
  photo: null,
  onPress: noop,
  number: 0
};

var GalleryThumbnail =
/*#__PURE__*/
function (_PureComponent) {
  _inherits(GalleryThumbnail, _PureComponent);

  function GalleryThumbnail() {
    _classCallCheck(this, GalleryThumbnail);

    return _possibleConstructorReturn(this, _getPrototypeOf(GalleryThumbnail).apply(this, arguments));
  }

  _createClass(GalleryThumbnail, [{
    key: "render",
    value: function render() {
      var _this$props = this.props,
          active = _this$props.active,
          photo = _this$props.photo,
          onPress = _this$props.onPress,
          number = _this$props.number;
      var className = classnames('thumbnail-button', active && 'active');
      return React.createElement("button", {
        type: "button",
        "aria-label": photo.caption,
        className: className,
        "data-photo-index": number,
        onClick: onPress,
        disabled: false,
        __source: {
          fileName: _jsxFileName$3,
          lineNumber: 54
        },
        __self: this
      }, React.createElement(Image, {
        alt: photo.caption,
        src: photo.thumbnail || photo.photo,
        className: "thumbnail",
        style: thumbnailStyle,
        __source: {
          fileName: _jsxFileName$3,
          lineNumber: 62
        },
        __self: this
      }));
    }
  }]);

  return GalleryThumbnail;
}(PureComponent);

GalleryThumbnail.propTypes = propTypes$3;
GalleryThumbnail.defaultProps = defaultProps$3;

var _jsxFileName$4 = "O:\\www\\www\\react\\react-bnb-gallery\\src\\components\\GalleryTogglePhotoList.js";
var propTypes$4 = forbidExtraProps({
  isOpened: PropTypes.bool,
  onPress: PropTypes.func,
  phrases: PropTypes.shape(defaultPhrases$1(defaultPhrases))
});
var defaultProps$4 = {
  isOpened: true,
  onPress: noop,
  phrases: defaultPhrases
};

var GalleryTogglePhotoList =
/*#__PURE__*/
function (_PureComponent) {
  _inherits(GalleryTogglePhotoList, _PureComponent);

  function GalleryTogglePhotoList() {
    _classCallCheck(this, GalleryTogglePhotoList);

    return _possibleConstructorReturn(this, _getPrototypeOf(GalleryTogglePhotoList).apply(this, arguments));
  }

  _createClass(GalleryTogglePhotoList, [{
    key: "render",
    value: function render() {
      var _this$props = this.props,
          isOpened = _this$props.isOpened,
          onPress = _this$props.onPress,
          _this$props$phrases = _this$props.phrases,
          showLabel = _this$props$phrases.showPhotoList,
          hideLabel = _this$props$phrases.hidePhotoList;
      var label = isOpened ? hideLabel : showLabel;
      var className = classnames('gallery-thumbnails--toggle', isOpened ? 'hide' : 'open');
      return React.createElement("button", {
        type: "button",
        className: className,
        onClick: onPress,
        __source: {
          fileName: _jsxFileName$4,
          lineNumber: 43
        },
        __self: this
      }, label);
    }
  }]);

  return GalleryTogglePhotoList;
}(PureComponent);

GalleryTogglePhotoList.propTypes = propTypes$4;
GalleryTogglePhotoList.defaultProps = defaultProps$4;

function calculateThumbnailsContainerDimension(total) {
  return THUMBNAIL_WIDTH * total + (THUMBNAIL_OFFSET * total - THUMBNAIL_OFFSET);
}

function calculateThumbnailsLeftScroll(current, total, bounding) {
  var half = bounding.width / 2 - THUMBNAIL_WIDTH / 2;
  var thumbnailsOffset = current * THUMBNAIL_WIDTH + current * THUMBNAIL_OFFSET - half;
  var calculatedScrollLeft = 0;

  if (thumbnailsOffset < 0) {
    return calculatedScrollLeft;
  }

  var thumbnailsPerRow = bounding.width / (THUMBNAIL_WIDTH + THUMBNAIL_OFFSET);
  var thumbnailsHalf = Math.round(thumbnailsPerRow / 2);
  var thumbnailsLeft = total - (current + 1);

  if (thumbnailsLeft < thumbnailsHalf) {
    calculatedScrollLeft = calculateThumbnailsContainerDimension(total) - bounding.width;
  } else {
    calculatedScrollLeft = thumbnailsOffset;
  }

  return -Math.abs(calculatedScrollLeft);
}

var _jsxFileName$5 = "O:\\www\\www\\react\\react-bnb-gallery\\src\\components\\GalleryCaption.js";
var propTypes$5 = forbidExtraProps({
  showThumbnails: PropTypes.bool,
  current: nonNegativeInteger,
  photos: PhotosShape,
  onPress: PropTypes.func,
  phrases: PropTypes.shape(defaultPhrases$1(defaultPhrases))
});
var defaultProps$5 = {
  showThumbnails: true,
  current: 0,
  photos: [],
  onPress: noop,
  phrases: defaultPhrases
};

var GalleryCaption =
/*#__PURE__*/
function (_PureComponent) {
  _inherits(GalleryCaption, _PureComponent);

  function GalleryCaption(props) {
    var _this;

    _classCallCheck(this, GalleryCaption);

    _this = _possibleConstructorReturn(this, _getPrototypeOf(GalleryCaption).call(this, props));
    var _this$props = _this.props,
        showThumbnails = _this$props.showThumbnails,
        photos = _this$props.photos;
    _this.state = {
      showThumbnails: showThumbnails
    };
    _this.thumbnailsWrapperRef = null;
    _this.thumbnailsListRef = null;
    _this.hasMoreThanOnePhoto = photos.length > 1;
    _this.onThumbnailPress = _this.onThumbnailPress.bind(_assertThisInitialized(_assertThisInitialized(_this)));
    _this.setGalleryFigcaptionRef = _this.setGalleryFigcaptionRef.bind(_assertThisInitialized(_assertThisInitialized(_this)));
    _this.setGalleryThubmanilsRef = _this.setGalleryThubmanilsRef.bind(_assertThisInitialized(_assertThisInitialized(_this)));
    _this.toggleThumbnails = _this.toggleThumbnails.bind(_assertThisInitialized(_assertThisInitialized(_this)));
    return _this;
  }

  _createClass(GalleryCaption, [{
    key: "componentDidUpdate",
    value: function componentDidUpdate(prevProps) {
      var current = this.props.current;

      if (current !== prevProps.current) {
        this.setThumbnailsWrapperScrollLeft(current);
      }
    }
  }, {
    key: "onThumbnailPress",
    value: function onThumbnailPress(event) {
      var _this$props2 = this.props,
          onPress = _this$props2.onPress,
          photos = _this$props2.photos;
      var index = parseInt(event.currentTarget.dataset.photoIndex, 10);

      if (index >= 0 && index <= photos.length - 1) {
        onPress(index);
      }
    }
  }, {
    key: "setThumbnailsWrapperScrollLeft",
    value: function setThumbnailsWrapperScrollLeft(current) {
      var photos = this.props.photos;
      var bounding = this.thumbnailsWrapperRef.getBoundingClientRect();
      var scrollLeft = calculateThumbnailsLeftScroll(current, photos.length, bounding);
      this.thumbnailsListRef.style.marginLeft = "".concat(scrollLeft, "px");
    }
  }, {
    key: "getPhotoByIndex",
    value: function getPhotoByIndex(index) {
      var photos = this.props.photos;
      return photos[index];
    }
  }, {
    key: "setGalleryFigcaptionRef",
    value: function setGalleryFigcaptionRef(element) {
      this.thumbnailsWrapperRef = element;
    }
  }, {
    key: "setGalleryThubmanilsRef",
    value: function setGalleryThubmanilsRef(element) {
      this.thumbnailsListRef = element;
    }
  }, {
    key: "toggleThumbnails",
    value: function toggleThumbnails() {
      this.setState(function (prevState) {
        return {
          showThumbnails: !prevState.showThumbnails
        };
      });
    }
  }, {
    key: "renderThumbnail",
    value: function renderThumbnail(photo, index, onPress) {
      var current = this.props.current;
      return React.createElement(GalleryThumbnail, {
        active: index === current,
        photo: photo,
        onPress: onPress,
        number: index,
        __source: {
          fileName: _jsxFileName$5,
          lineNumber: 108
        },
        __self: this
      });
    }
  }, {
    key: "render",
    value: function render() {
      var _this2 = this;

      var _this$props3 = this.props,
          current = _this$props3.current,
          photos = _this$props3.photos,
          phrases = _this$props3.phrases;
      var showThumbnails = this.state.showThumbnails;
      var className = classnames('gallery-figcaption', !showThumbnails && 'hide');
      var currentPhoto = this.getPhotoByIndex(current);
      var captionThumbnailsWrapperWidth = calculateThumbnailsContainerDimension(photos.length);
      return React.createElement("figcaption", {
        className: className,
        __source: {
          fileName: _jsxFileName$5,
          lineNumber: 137
        },
        __self: this
      }, React.createElement("div", {
        className: "gallery-figcaption--content",
        __source: {
          fileName: _jsxFileName$5,
          lineNumber: 138
        },
        __self: this
      }, React.createElement("div", {
        className: "gallery-figcaption--inner",
        __source: {
          fileName: _jsxFileName$5,
          lineNumber: 139
        },
        __self: this
      }, React.createElement("div", {
        className: "gallery-figcaption--info",
        __source: {
          fileName: _jsxFileName$5,
          lineNumber: 140
        },
        __self: this
      }, React.createElement("div", {
        className: "caption-left",
        __source: {
          fileName: _jsxFileName$5,
          lineNumber: 141
        },
        __self: this
      }, currentPhoto.caption && React.createElement("h3", {
        className: "photo-caption",
        __source: {
          fileName: _jsxFileName$5,
          lineNumber: 143
        },
        __self: this
      }, currentPhoto.caption), currentPhoto.subcaption && React.createElement("p", {
        className: "photo-subcaption",
        __source: {
          fileName: _jsxFileName$5,
          lineNumber: 148
        },
        __self: this
      }, currentPhoto.subcaption)), this.hasMoreThanOnePhoto && React.createElement("div", {
        className: "caption-right",
        __source: {
          fileName: _jsxFileName$5,
          lineNumber: 154
        },
        __self: this
      }, React.createElement(GalleryTogglePhotoList, {
        phrases: phrases,
        isOpened: showThumbnails,
        onPress: this.toggleThumbnails,
        __source: {
          fileName: _jsxFileName$5,
          lineNumber: 155
        },
        __self: this
      }))), this.hasMoreThanOnePhoto && React.createElement("div", {
        className: "gallery-figcaption--thumbnails",
        "aria-hidden": false,
        ref: this.setGalleryFigcaptionRef,
        __source: {
          fileName: _jsxFileName$5,
          lineNumber: 164
        },
        __self: this
      }, React.createElement("div", {
        className: "caption-thumbnails",
        style: {
          width: captionThumbnailsWrapperWidth
        },
        __source: {
          fileName: _jsxFileName$5,
          lineNumber: 169
        },
        __self: this
      }, React.createElement("ul", {
        className: "thumbnails-list",
        ref: this.setGalleryThubmanilsRef,
        __source: {
          fileName: _jsxFileName$5,
          lineNumber: 173
        },
        __self: this
      }, photos.map(function (photo, index) {
        return React.createElement("li", {
          key: photo.photo,
          __source: {
            fileName: _jsxFileName$5,
            lineNumber: 178
          },
          __self: this
        }, _this2.renderThumbnail(photo, index, _this2.onThumbnailPress));
      })))))));
    }
  }]);

  return GalleryCaption;
}(PureComponent);

GalleryCaption.propTypes = propTypes$5;
GalleryCaption.defaultProps = defaultProps$5;

var _jsxFileName$6 = "O:\\www\\www\\react\\react-bnb-gallery\\src\\components\\GalleryControl.js";
var svgAttributes = {
  viewBox: '0 0 18 18',
  role: 'presentation',
  focusable: false,
  'aria-hidden': true
};
var controlStyle = {
  height: '2.8em',
  width: '2.8em',
  fill: NORMAL_COLOR
};
var controlStyleLight = {
  fill: INVERSE_COLOR
};
var propTypes$6 = forbidExtraProps({
  arrow: PropTypes.string,
  onPress: PropTypes.func,
  label: PropTypes.string,
  className: PropTypes.string,
  disabled: PropTypes.bool,
  light: PropTypes.bool
});
var defaultProps$6 = {
  arrow: null,
  onPress: noop,
  label: '',
  className: null,
  disabled: false,
  light: false
};

var GalleryControl =
/*#__PURE__*/
function (_React$PureComponent) {
  _inherits(GalleryControl, _React$PureComponent);

  function GalleryControl(props) {
    var _this;

    _classCallCheck(this, GalleryControl);

    _this = _possibleConstructorReturn(this, _getPrototypeOf(GalleryControl).call(this, props));
    _this.onButtonPress = _this.onButtonPress.bind(_assertThisInitialized(_assertThisInitialized(_this)));
    return _this;
  }

  _createClass(GalleryControl, [{
    key: "onButtonPress",
    value: function onButtonPress() {
      var onPress = this.props.onPress;
      onPress();
      return false;
    }
  }, {
    key: "render",
    value: function render() {
      var _this$props = this.props,
          arrow = _this$props.arrow,
          label = _this$props.label,
          className = _this$props.className,
          disabled = _this$props.disabled,
          light = _this$props.light;
      return React.createElement("button", {
        type: "button",
        className: classnames('gallery-control', className),
        onClick: this.onButtonPress,
        disabled: disabled,
        "aria-label": label,
        __source: {
          fileName: _jsxFileName$6,
          lineNumber: 71
        },
        __self: this
      }, React.createElement("svg", _extends({}, svgAttributes, {
        style: _objectSpread({}, controlStyle, light && controlStyleLight),
        __source: {
          fileName: _jsxFileName$6,
          lineNumber: 78
        },
        __self: this
      }), React.createElement("path", {
        d: arrow,
        fillRule: "evenodd",
        __source: {
          fileName: _jsxFileName$6,
          lineNumber: 85
        },
        __self: this
      })));
    }
  }]);

  return GalleryControl;
}(React.PureComponent);

GalleryControl.propTypes = propTypes$6;
GalleryControl.defaultProps = defaultProps$6;

var _jsxFileName$7 = "O:\\www\\www\\react\\react-bnb-gallery\\src\\components\\GalleryPrevButton.js";
var PREV_ARROW = 'm13.7 16.29a1 1 0 1 1 -1.42 1.41l-8-8a1 1 0 0 1 0-1.41l8-8a1 1 0 1 1 1.42 1.41l-7.29 7.29z';
var propTypes$7 = forbidExtraProps({
  onPress: PropTypes.func,
  disabled: PropTypes.bool,
  light: PropTypes.bool
});
var defaultProps$7 = {
  onPress: noop,
  disabled: false,
  light: false
};

var GalleryPrevButton =
/*#__PURE__*/
function (_PureComponent) {
  _inherits(GalleryPrevButton, _PureComponent);

  function GalleryPrevButton() {
    _classCallCheck(this, GalleryPrevButton);

    return _possibleConstructorReturn(this, _getPrototypeOf(GalleryPrevButton).apply(this, arguments));
  }

  _createClass(GalleryPrevButton, [{
    key: "render",
    value: function render() {
      var _this$props = this.props,
          onPress = _this$props.onPress,
          disabled = _this$props.disabled,
          light = _this$props.light;
      return React.createElement(GalleryControl, {
        className: "gallery-control--prev",
        onPress: onPress,
        arrow: PREV_ARROW,
        disabled: disabled,
        light: light,
        __source: {
          fileName: _jsxFileName$7,
          lineNumber: 32
        },
        __self: this
      });
    }
  }]);

  return GalleryPrevButton;
}(PureComponent);

GalleryPrevButton.propTypes = propTypes$7;
GalleryPrevButton.defaultProps = defaultProps$7;

var _jsxFileName$8 = "O:\\www\\www\\react\\react-bnb-gallery\\src\\components\\GalleryNextButton.js";
var NEXT_ARROW = 'm4.29 1.71a1 1 0 1 1 1.42-1.41l8 8a1 1 0 0 1 0 1.41l-8 8a1 1 0 1 1 -1.42-1.41l7.29-7.29z';
var propTypes$8 = forbidExtraProps({
  onPress: PropTypes.func,
  disabled: PropTypes.bool,
  light: PropTypes.bool
});
var defaultProps$8 = {
  onPress: noop,
  disabled: false,
  light: false
};

var GalleryNextButton =
/*#__PURE__*/
function (_PureComponent) {
  _inherits(GalleryNextButton, _PureComponent);

  function GalleryNextButton() {
    _classCallCheck(this, GalleryNextButton);

    return _possibleConstructorReturn(this, _getPrototypeOf(GalleryNextButton).apply(this, arguments));
  }

  _createClass(GalleryNextButton, [{
    key: "render",
    value: function render() {
      var _this$props = this.props,
          onPress = _this$props.onPress,
          disabled = _this$props.disabled,
          light = _this$props.light;
      return React.createElement(GalleryControl, {
        className: "gallery-control--next",
        onPress: onPress,
        arrow: NEXT_ARROW,
        disabled: disabled,
        light: light,
        __source: {
          fileName: _jsxFileName$8,
          lineNumber: 32
        },
        __self: this
      });
    }
  }]);

  return GalleryNextButton;
}(PureComponent);

GalleryNextButton.propTypes = propTypes$8;
GalleryNextButton.defaultProps = defaultProps$8;

var _jsxFileName$9 = "O:\\www\\www\\react\\react-bnb-gallery\\src\\components\\Gallery.js";

var propTypes$9 = _objectSpread({}, galleryPropTypes);

var defaultProps$9 = _objectSpread({}, galleryDefaultProps);

var Gallery =
/*#__PURE__*/
function (_PureComponent) {
  _inherits(Gallery, _PureComponent);

  function Gallery(props) {
    var _this;

    _classCallCheck(this, Gallery);

    _this = _possibleConstructorReturn(this, _getPrototypeOf(Gallery).call(this, props));
    var _this$props = _this.props,
        activePhotoIndex = _this$props.activePhotoIndex,
        photos = _this$props.photos,
        wrap = _this$props.wrap;
    _this.state = {
      activePhotoIndex: activePhotoIndex,
      hidePrevButton: wrap && activePhotoIndex === 0,
      hideNextButton: wrap && activePhotoIndex === photos.length - 1,
      controlsDisabled: true
    };
    _this.lastPreloadIndex = 0;
    _this.preloadedPhotos = [];
    _this.hasPhotos = photos.length > 0;
    _this.hasMoreThanOnePhoto = photos.length > 1;
    _this.move = _this.move.bind(_assertThisInitialized(_assertThisInitialized(_this)));
    _this.prev = _this.prev.bind(_assertThisInitialized(_assertThisInitialized(_this)));
    _this.next = _this.next.bind(_assertThisInitialized(_assertThisInitialized(_this)));
    _this.onPhotoLoad = _this.onPhotoLoad.bind(_assertThisInitialized(_assertThisInitialized(_this)));
    _this.onPhotoError = _this.onPhotoError.bind(_assertThisInitialized(_assertThisInitialized(_this)));
    _this.onPhotoPress = _this.onPhotoPress.bind(_assertThisInitialized(_assertThisInitialized(_this)));
    _this.onThumbnailPress = _this.onThumbnailPress.bind(_assertThisInitialized(_assertThisInitialized(_this)));
    _this.onPrevButtonPress = _this.onPrevButtonPress.bind(_assertThisInitialized(_assertThisInitialized(_this)));
    _this.onNextButtonPress = _this.onNextButtonPress.bind(_assertThisInitialized(_assertThisInitialized(_this)));
    return _this;
  }

  _createClass(Gallery, [{
    key: "onNextButtonPress",
    value: function onNextButtonPress() {
      var nextButtonPressed = this.props.nextButtonPressed;
      this.next();
      nextButtonPressed();
    }
  }, {
    key: "onPrevButtonPress",
    value: function onPrevButtonPress() {
      var prevButtonPressed = this.props.prevButtonPressed;
      this.prev();
      prevButtonPressed();
    }
  }, {
    key: "onPhotoLoad",
    value: function onPhotoLoad() {
      return this.setState({
        controlsDisabled: false
      });
    }
  }, {
    key: "onPhotoError",
    value: function onPhotoError() {
      return this.setState({
        controlsDisabled: false
      });
    }
  }, {
    key: "onPhotoPress",
    value: function onPhotoPress() {
      var activePhotoPressed = this.props.activePhotoPressed;
      this.move(DIRECTION_NEXT);
      activePhotoPressed();
    }
  }, {
    key: "onThumbnailPress",
    value: function onThumbnailPress(index) {
      this.to(index);
    }
  }, {
    key: "getPhotoByIndex",
    value: function getPhotoByIndex(index) {
      var photos = this.props.photos;
      return photos[index];
    }
  }, {
    key: "getItemByDirection",
    value: function getItemByDirection(direction, activeIndex) {
      var _this$props2 = this.props,
          photos = _this$props2.photos,
          wrap = _this$props2.wrap;
      var isNextDirection = direction === DIRECTION_NEXT;
      var isPrevDirection = direction === DIRECTION_PREV;
      var lastItemIndex = photos.length - 1;
      var isGoingToWrap = isPrevDirection && activeIndex === 0 || isNextDirection && activeIndex === lastItemIndex;

      if (isGoingToWrap && wrap) {
        return activeIndex;
      }

      var delta = isPrevDirection ? -1 : 1;
      var itemIndex = (activeIndex + delta) % photos.length;
      return itemIndex === -1 ? photos.length - 1 : itemIndex;
    }
  }, {
    key: "to",
    value: function to(index) {
      var photos = this.props.photos;
      var activePhotoIndex = this.state.activePhotoIndex;

      if (index > photos.length - 1 || index < 0 || activePhotoIndex === index) {
        return; // nothing to do
      }

      var direction = index > activePhotoIndex ? DIRECTION_NEXT : DIRECTION_PREV;
      this.move(direction, index);
    }
  }, {
    key: "move",
    value: function move(direction) {
      var index = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : false;
      var activePhotoIndex = this.state.activePhotoIndex;
      var nextElementIndex = index !== false ? index : this.getItemByDirection(direction, activePhotoIndex);
      this.wrapCheck(direction, nextElementIndex);
      this.setState({
        activePhotoIndex: nextElementIndex
      });
    }
  }, {
    key: "next",
    value: function next() {
      return this.move(DIRECTION_NEXT);
    }
  }, {
    key: "prev",
    value: function prev() {
      return this.move(DIRECTION_PREV);
    }
  }, {
    key: "wrapCheck",
    value: function wrapCheck(direction, nextElementIndex) {
      var _this$props3 = this.props,
          photos = _this$props3.photos,
          wrap = _this$props3.wrap;

      if (wrap) {
        this.setState({
          hideNextButton: nextElementIndex === photos.length - 1,
          hidePrevButton: nextElementIndex === 0
        });
      }
    }
  }, {
    key: "renderControls",
    value: function renderControls() {
      var light = this.props.light;
      var _this$state = this.state,
          hidePrevButton = _this$state.hidePrevButton,
          hideNextButton = _this$state.hideNextButton,
          controlsDisabled = _this$state.controlsDisabled;
      var controls = [];

      if (this.hasMoreThanOnePhoto) {
        // previous control
        if (!hidePrevButton) {
          controls.push(React.createElement(GalleryPrevButton, {
            key: ".prevControl",
            disabled: controlsDisabled,
            onPress: this.onPrevButtonPress,
            light: light,
            __source: {
              fileName: _jsxFileName$9,
              lineNumber: 175
            },
            __self: this
          }));
        } // next control


        if (!hideNextButton) {
          controls.push(React.createElement(GalleryNextButton, {
            key: ".nextControl",
            disabled: controlsDisabled,
            onPress: this.onNextButtonPress,
            light: light,
            __source: {
              fileName: _jsxFileName$9,
              lineNumber: 187
            },
            __self: this
          }));
        }
      }

      return controls;
    }
  }, {
    key: "renderPreloadPhotos",
    value: function renderPreloadPhotos() {
      var photos = this.props.photos;
      return photos.map(function (photo) {
        return React.createElement("img", {
          key: photo.photo,
          alt: photo.photo,
          src: photo.photo,
          __source: {
            fileName: _jsxFileName$9,
            lineNumber: 204
          },
          __self: this
        });
      });
    }
  }, {
    key: "render",
    value: function render() {
      var _this$props4 = this.props,
          photos = _this$props4.photos,
          showThumbnails = _this$props4.showThumbnails,
          phrases = _this$props4.phrases;
      var emptyMessage = phrases.noPhotosProvided;
      var activePhotoIndex = this.state.activePhotoIndex; // preload photos

      var galleryModalPreloadPhotos = this.renderPreloadPhotos();
      var controls = this.renderControls();
      var current = this.getPhotoByIndex(activePhotoIndex);
      return React.createElement("div", {
        className: "gallery",
        __source: {
          fileName: _jsxFileName$9,
          lineNumber: 235
        },
        __self: this
      }, React.createElement("div", {
        className: "gallery-modal--preload",
        __source: {
          fileName: _jsxFileName$9,
          lineNumber: 236
        },
        __self: this
      }, galleryModalPreloadPhotos), React.createElement("div", {
        className: "gallery-main",
        __source: {
          fileName: _jsxFileName$9,
          lineNumber: 239
        },
        __self: this
      }, controls, React.createElement("div", {
        className: "gallery-photos",
        __source: {
          fileName: _jsxFileName$9,
          lineNumber: 241
        },
        __self: this
      }, this.hasPhotos ? React.createElement("div", {
        className: "gallery-photo",
        __source: {
          fileName: _jsxFileName$9,
          lineNumber: 243
        },
        __self: this
      }, React.createElement("div", {
        className: "gallery-photo--current",
        __source: {
          fileName: _jsxFileName$9,
          lineNumber: 244
        },
        __self: this
      }, React.createElement(GalleryPhoto, {
        photo: current,
        onLoad: this.onPhotoLoad,
        onError: this.onPhotoError,
        onPress: this.onPhotoPress,
        __source: {
          fileName: _jsxFileName$9,
          lineNumber: 245
        },
        __self: this
      }))) : React.createElement("div", {
        className: "gallery-empty",
        __source: {
          fileName: _jsxFileName$9,
          lineNumber: 254
        },
        __self: this
      }, emptyMessage))), showThumbnails && current && React.createElement(GalleryCaption, {
        phrases: phrases,
        current: activePhotoIndex,
        photos: photos,
        onPress: this.onThumbnailPress,
        __source: {
          fileName: _jsxFileName$9,
          lineNumber: 261
        },
        __self: this
      }));
    }
  }]);

  return Gallery;
}(PureComponent);

Gallery.propTypes = propTypes$9;
Gallery.defaultProps = defaultProps$9;

var _jsxFileName$a = "O:\\www\\www\\react\\react-bnb-gallery\\src\\components\\GalleryCloseButton.js";
var CLOSE_PATH = 'm23.25 24c-.19 0-.38-.07-.53-.22l-10.72-10.72-10.72 10.72c-.29.29-.77.29-1.06 0s-.29-.77 0-1.06l10.72-10.72-10.72-10.72c-.29-.29-.29-.77 0-1.06s.77-.29 1.06 0l10.72 10.72 10.72-10.72c.29-.29.77-.29 1.06 0s .29.77 0 1.06l-10.72 10.72 10.72 10.72c.29.29.29.77 0 1.06-.15.15-.34.22-.53.22';
var buttonStyle = {
  height: '2em',
  width: '2em',
  display: 'block',
  fill: NORMAL_COLOR
};
var buttonStyleLight = {
  fill: INVERSE_COLOR
};
var buttonProps = {
  viewBox: '0 0 24 24',
  role: 'img',
  focusable: false
};
var propTypes$a = forbidExtraProps({
  onPress: PropTypes.func,
  light: PropTypes.bool
});
var defaultProps$a = {
  onPress: noop,
  light: false
};

var GalleryCloseButton = function GalleryCloseButton(_ref) {
  var onPress = _ref.onPress,
      light = _ref.light;
  return React.createElement("button", {
    onClick: onPress,
    className: "gallery-close",
    type: "button",
    "aria-busy": false,
    __source: {
      fileName: _jsxFileName$a,
      lineNumber: 45
    },
    __self: this
  }, React.createElement("svg", _extends({}, buttonProps, {
    style: _objectSpread({}, buttonStyle, light && buttonStyleLight),
    __source: {
      fileName: _jsxFileName$a,
      lineNumber: 51
    },
    __self: this
  }), React.createElement("path", {
    d: CLOSE_PATH,
    fillRule: "evenodd",
    __source: {
      fileName: _jsxFileName$a,
      lineNumber: 58
    },
    __self: this
  })));
};

GalleryCloseButton.propTypes = propTypes$a;
GalleryCloseButton.defaultProps = defaultProps$a;

var MIN = 0;
var MAX = 1; // eslint-disable-next-line consistent-return

function opacityValidation(props, propName) {
  if (props[propName] < MIN || props[propName] > MAX) {
    return new Error('Invalid value for opacity');
  }
}

function processPhoto(photo, index) {
  return _extends({}, defaultPhotoProps, typeof photo === 'string' ? {
    number: index + 1,
    photo: photo
  } : _objectSpread({}, photo, {
    number: index + 1
  }));
}
function getPhotos(photos) {
  return photos.map(processPhoto);
}

function styleInject(css, ref) {
  if ( ref === void 0 ) ref = {};
  var insertAt = ref.insertAt;

  if (!css || typeof document === 'undefined') { return; }

  var head = document.head || document.getElementsByTagName('head')[0];
  var style = document.createElement('style');
  style.type = 'text/css';

  if (insertAt === 'top') {
    if (head.firstChild) {
      head.insertBefore(style, head.firstChild);
    } else {
      head.appendChild(style);
    }
  } else {
    head.appendChild(style);
  }

  if (style.styleSheet) {
    style.styleSheet.cssText = css;
  } else {
    style.appendChild(document.createTextNode(css));
  }
}

var css = ".gallery-modal {\r\n  position: fixed;\r\n  top: 0px;\r\n  right: 0px;\r\n  bottom: 0px;\r\n  left: 0px;\r\n  z-index: 2000;\r\n}\r\n\r\n.gallery-modal .gallery-modal--overlay {\r\n  background-color: rgba(0,0,0,1);\r\n  position: absolute;\r\n  left: 0;\r\n  top: 0;\r\n  width: 100%;\r\n  height: 100%;\r\n  right: 0;\r\n  bottom: 0;\r\n  z-index: 0;\r\n}\r\n\r\n.gallery-modal--preload {\r\n  display: none;\r\n}\r\n\r\n.gallery-modal--container {\r\n  position: fixed;\r\n  z-index: 2000;\r\n  top: 0;\r\n  right: 0;\r\n  bottom: 0;\r\n  left: 0;\r\n  overflow-y: auto;\r\n  -webkit-transform: translate3d(0, 0, 0);\r\n  z-index: 1;\r\n}\r\n\r\n.gallery-modal--table {\r\n  display: table;\r\n  table-layout: fixed;\r\n  height: 100%;\r\n  width: 100%;\r\n}\r\n\r\n.gallery-modal--cell {\r\n  display: table-cell;\r\n  height: 100%;\r\n  width: 100%;\r\n  vertical-align: middle;\r\n}\r\n\r\n.gallery-modal--content {\r\n  position: relative;\r\n  overflow: hidden;\r\n  -webkit-border-radius: 2px;\r\n  -moz-border-radius: 2px;\r\n  border-radius: 2px;\r\n  margin-left: auto;\r\n  margin-right: auto;\r\n  background-color: transparent;\r\n  height: 100%;\r\n  max-width: 100%;\r\n}\r\n\r\n.gallery-content {\r\n  display: table;\r\n  width: 100%;\r\n  height: 100%;\r\n}\r\n\r\n.gallery-top {\r\n  display: table-row;\r\n}\r\n\r\n.gallery-top--inner {\r\n  padding: 60px 15px 20px;\r\n}\r\n\r\n.gallery {\r\n  display: table-row;\r\n  height: 100%;\r\n  font-size: 1rem;\r\n}\r\n\r\n.gallery .gallery-main {\r\n  position: relative;\r\n  z-index: 9;\r\n}\r\n\r\n.gallery .gallery-photos {\r\n  width: 100%;\r\n  max-width: 105vh;\r\n  margin: 0px auto;\r\n}\r\n\r\n.gallery .gallery-photo {\r\n  position: relative;\r\n  width: 100%;\r\n  height: 0px;\r\n  padding-bottom: 75%;\r\n}\r\n\r\n.gallery .gallery-photos .gallery-empty {\r\n  color: #ffffff;\r\n  font-size: 1.45rem;\r\n  display: block;\r\n  text-align: center;\r\n  padding: 25% 0;\r\n}\r\n\r\n.gallery .gallery-photo--current {\r\n  position: absolute;\r\n  top: 0px;\r\n  right: 0px;\r\n  bottom: 0px;\r\n  left: 0px;\r\n  width: 100%;\r\n  height: 100%;\r\n}\r\n\r\n.gallery ul.gallery-images--ul {\r\n  position: relative;\r\n  height: 100%;\r\n  list-style-type: none;\r\n  margin: 0;\r\n  padding: 0;\r\n}\r\n\r\n.gallery ul.gallery-images--ul li.gallery-media-photo {\r\n  opacity: 1;\r\n  -ms-filter: 'progid:DXImageTransform.Microsoft.Alpha(Opacity=100)';\r\n  z-index: 2;\r\n}\r\n\r\n.gallery ul.gallery-images--ul li.gallery-media-photo,\r\n.gallery ul.gallery-images--ul li.gallery-media-photo div.picture {\r\n  background-color: transparent;\r\n  cursor: pointer;\r\n  position: absolute;\r\n  width: 100%;\r\n  height: 100%;\r\n  overflow: hidden;\r\n}\r\n\r\n.gallery ul.gallery-images--ul li.gallery-media-photo button.photo-button {\r\n  display: block;\r\n  background: none;\r\n  border-width: 0;\r\n  cursor: pointer;\r\n  padding: 0 !important;\r\n  -webkit-appearance: button;\r\n}\r\n\r\n.gallery ul.gallery-images--ul li.gallery-media-photo button.photo-button:focus,\r\n.gallery ul.gallery-images--ul li.gallery-media-photo button.photo-button:active {\r\n  outline: none;\r\n  border-width: 0;\r\n}\r\n\r\n.gallery ul.gallery-images--ul li.gallery-media-photo div.picture img.photo {\r\n  max-height: 100%;\r\n  max-width: 100%;\r\n  position: absolute;\r\n  top: 50%;\r\n  right: 0;\r\n  bottom: 0;\r\n  left: 0;\r\n  margin: 0 auto;\r\n  z-index: 0;\r\n  -webkit-transform: translateY(-50%);\r\n  transform: translateY(-50%);\r\n}\r\n\r\n.gallery .gallery-control {\r\n  cursor: pointer;\r\n  position: absolute;\r\n  top: 0px;\r\n  height: 100%;\r\n  z-index: 3;\r\n  font-size: 0.5em;\r\n  width: 40px;\r\n  background: none;\r\n  border-width: 0px;\r\n  border-style: initial;\r\n  border-color: initial;\r\n  border-image: initial;\r\n}\r\n\r\n.gallery .gallery-control--prev {\r\n  left: 0;\r\n}\r\n\r\n.gallery .gallery-control--next {\r\n  right: 0;\r\n}\r\n\r\n.gallery-figcaption {\r\n  background-image: none;\r\n  background-color: transparent;\r\n  position: relative;\r\n  right: 0;\r\n  bottom: 0;\r\n  left: 0;\r\n  z-index: 2;\r\n  text-align: center;\r\n  color: rgb(255, 255, 255);\r\n  overflow: hidden;\r\n  border-radius: 0;\r\n}\r\n\r\n.mode-light .gallery-figcaption {\r\n  color: #111111;\r\n}\r\n\r\n.gallery-figcaption--content {\r\n  max-width: 105vh;\r\n  margin-left: auto;\r\n  margin-right: auto;\r\n  overflow: hidden;\r\n}\r\n\r\n.gallery-figcaption--inner {\r\n  margin-left: auto;\r\n  margin-right: auto;\r\n  position: relative;\r\n  transform: translateY(0);\r\n  transition: -ms-transform 0.2s ease-out 0s, -webkit-transform 0.2s ease-out 0s, transform 0.2s ease-out 0s;\r\n}\r\n\r\n.hide .gallery-figcaption--inner {\r\n  transform: translateY(70px);\r\n}\r\n\r\n.gallery-figcaption--info {\r\n  display: table;\r\n  width: 100%;\r\n}\r\n\r\n.gallery-figcaption--info .caption-left {\r\n  text-align: left;\r\n  vertical-align: middle;\r\n  display: table-cell;\r\n}\r\n\r\n.gallery-figcaption--info .caption-right {\r\n  text-align: right;\r\n  vertical-align: middle;\r\n  display: table-cell;\r\n}\r\n\r\n.gallery-figcaption--info .photo-caption,\r\n.gallery-figcaption--info .photo-subcaption {\r\n  margin: 0;\r\n  line-height: 1.31;\r\n}\r\n\r\n.gallery-figcaption--info .photo-caption {\r\n  font-weight: 400;\r\n  font-size: 1rem;\r\n}\r\n\r\n.gallery-figcaption--info .photo-subcaption {\r\n  color: rgba(255,255,255,0.65);\r\n  font-size: 0.851rem;\r\n}\r\n\r\n.mode-light .gallery-figcaption--info .photo-subcaption {\r\n  color: rgba(0,0,0,0.65);\r\n}\r\n\r\n.gallery-figcaption--thumbnails {\r\n  position: relative;\r\n  height: 67px;\r\n  overflow: hidden;\r\n}\r\n\r\n.gallery-figcaption--thumbnails .caption-thumbnails {\r\n  position: absolute;\r\n}\r\n\r\n.gallery-figcaption--thumbnails .caption-thumbnails ul.thumbnails-list {\r\n  position: absolute;\r\n  list-style-type: none;\r\n  margin: 0;\r\n  padding: 0;\r\n  left: 0;\r\n  -webkit-transition: margin 0.3s ease-out 0s;\r\n  transition: margin 0.3s ease-out 0s;\r\n}\r\n\r\n.gallery-figcaption--thumbnails .caption-thumbnails ul.thumbnails-list::before,\r\n.gallery-figcaption--thumbnails .caption-thumbnails ul.thumbnails-list::after {\r\n  content: \"\";\r\n  display: table;\r\n}\r\n\r\n.gallery-figcaption--thumbnails .caption-thumbnails ul.thumbnails-list::after {\r\n  clear: both;\r\n}\r\n\r\n.gallery-figcaption--thumbnails .caption-thumbnails ul.thumbnails-list li {\r\n  float: left;\r\n  background-color: rgb(0, 0, 0);\r\n}\r\n\r\n.mode-light .gallery-figcaption--thumbnails .caption-thumbnails ul.thumbnails-list li {\r\n  background-color: rgb(255, 255, 255);\r\n}\r\n\r\n.gallery-figcaption--thumbnails .caption-thumbnails ul.thumbnails-list li + li {\r\n  margin-left: 10px;\r\n}\r\n\r\n.gallery-figcaption--thumbnails .caption-thumbnails ul.thumbnails-list li button.thumbnail-button {\r\n  cursor: pointer;\r\n  position: relative;\r\n  display: block;\r\n  vertical-align: bottom;\r\n  overflow: hidden;\r\n  background-color: #bbb;\r\n  background: transparent;\r\n  border: 0;\r\n  margin: 0;\r\n  padding: 0;\r\n}\r\n\r\n.gallery-figcaption--thumbnails .caption-thumbnails ul.thumbnails-list li button.thumbnail-button div.loading-spinner {\r\n  position: absolute;\r\n  z-index: 1001;\r\n  width: 2em;\r\n  height: 2em;\r\n  top: 50%;\r\n  left: 50%;\r\n  margin: -1em 0 0 -1em;\r\n  border-top-width: 0.15em ;\r\n  border-right-width: 0.15em;\r\n  border-bottom-width: 0.15em;\r\n  border-left-width: 0.15em;\r\n  -webkit-transform: translate(-50%,-50%);\r\n  transform: translate(-50%,-50%);\r\n}\r\n\r\n.gallery-figcaption--thumbnails .caption-thumbnails ul.thumbnails-list li button.thumbnail-button.active {\r\n  cursor: default;\r\n}\r\n\r\n.gallery-figcaption--thumbnails .caption-thumbnails ul.thumbnails-list li img.thumbnail {\r\n  width: 100px;\r\n  height: 67px;\r\n  display: block;\r\n  border-width: 0;\r\n  position: relative;\r\n  z-index: 1;\r\n  opacity: 0.3;\r\n  -ms-filter: \"progid:DXImageTransform.Microsoft.Alpha(Opacity=30)\";\r\n  filter: alpha(opacity=30);\r\n  -webkit-backface-visibility: hidden;\r\n  -moz-backface-visibility: hidden;\r\n  backface-visibility: hidden;\r\n  -webkit-transition: 0.1s ease opacity;\r\n  transition: 0.1s ease opacity;\r\n}\r\n\r\n.gallery-figcaption--thumbnails .caption-thumbnails ul.thumbnails-list li button.thumbnail-button.active img,\r\n.gallery-figcaption--thumbnails .caption-thumbnails ul.thumbnails-list li button.thumbnail-button:hover img {\r\n  opacity: 1;\r\n  -ms-filter: \"progid:DXImageTransform.Microsoft.Alpha(Opacity=100)\";\r\n  filter: alpha(opacity=100);\r\n}\r\n\r\n.gallery-figcaption .gallery-thumbnails--toggle {\r\n  cursor: pointer;\r\n  background: transparent;\r\n  border: 0;\r\n  margin: 0;\r\n  padding: 0;\r\n  vertical-align: bottom;\r\n  display: inline-block;\r\n  color: inherit;\r\n  font-weight: inherit;\r\n  position: relative;\r\n}\r\n\r\n.gallery-figcaption .gallery-thumbnails--toggle::after {\r\n  content: \"\";\r\n  display: inline-block;\r\n  vertical-align: middle;\r\n  margin-left: 0.181rem;\r\n}\r\n\r\n.gallery-figcaption .gallery-thumbnails--toggle.hide::after {\r\n  border-top: 6px solid #ffffff;\r\n  border-left: 6px solid transparent;\r\n  border-right: 6px solid transparent;\r\n}\r\n\r\n.mode-light .gallery-figcaption .gallery-thumbnails--toggle.hide::after {\r\n  border-top-color: #111111;\r\n}\r\n\r\n.gallery-figcaption .gallery-thumbnails--toggle.open::after {\r\n  border-bottom: 6px solid #ffffff;\r\n  border-left: 6px solid transparent;\r\n  border-right: 6px solid transparent;\r\n}\r\n\r\n.mode-light .gallery-figcaption .gallery-thumbnails--toggle.open::after {\r\n  border-bottom-color: #111111;\r\n}\r\n\r\n.gallery-modal--close {\r\n  position: absolute;\r\n  z-index: 10;\r\n  right: 0px;\r\n  top: 0px;\r\n  padding: 15px;\r\n}\r\n\r\n.gallery-modal--close button.gallery-close {\r\n  cursor: pointer;\r\n  background: none transparent;\r\n  color: 'buttontext';\r\n  border: 0;\r\n  display: block;\r\n  padding: 0.5rem;\r\n}\r\n\r\ndiv.loading-spinner,\r\ndiv.loading-spinner::after,\r\n.gallery-figcaption--thumbnails .caption-thumbnails ul.thumbnails-list li button.thumbnail-button::before,\r\n.gallery-figcaption--thumbnails .caption-thumbnails ul.thumbnails-list li button.thumbnail-button::after {\r\n  border-radius: 50%;\r\n}\r\n\r\ndiv.loading-spinner,\r\ndiv.loading-spinner::after {\r\n  width: 6em;\r\n  height: 6em;\r\n}\r\n\r\ndiv.loading-spinner,\r\n.gallery-figcaption--thumbnails .caption-thumbnails ul.thumbnails-list li button.thumbnail-button::after {\r\n  font-size: 10px;\r\n  text-indent: -9999em;\r\n  -webkit-transform: translateZ(0);\r\n  -ms-transform: translateZ(0);\r\n  transform: translateZ(0);\r\n  -webkit-animation: GalleryLoadingSpinner 1.1s infinite linear;\r\n  animation: GalleryLoadingSpinner 1.1s infinite linear;\r\n}\r\n\r\ndiv.loading-spinner {\r\n  margin: 60px auto;\r\n  position: relative;\r\n  border-top: 0.35em solid rgba(255, 255, 255, 0.2);\r\n  border-right: 0.35em solid rgba(255, 255, 255, 0.2);\r\n  border-bottom: 0.35em solid rgba(255, 255, 255, 0.2);\r\n  border-left: 0.35em solid #ffffff;\r\n}\r\n\r\n.mode-light div.loading-spinner {\r\n  border-top-color: rgba(0, 0, 0, 0.1);\r\n  border-right-color: rgba(0, 0, 0, 0.1);\r\n  border-bottom-color: rgba(0, 0, 0, 0.1);\r\n  border-left-color: #111111;\r\n}\r\n\r\n@-webkit-keyframes GalleryLoadingSpinner {\r\n  0% {\r\n    -webkit-transform: rotate(0deg);\r\n    transform: rotate(0deg);\r\n  }\r\n  100% {\r\n    -webkit-transform: rotate(360deg);\r\n    transform: rotate(360deg);\r\n  }\r\n}\r\n\r\n@keyframes GalleryLoadingSpinner {\r\n  0% {\r\n    -webkit-transform: rotate(0deg);\r\n    transform: rotate(0deg);\r\n  }\r\n  100% {\r\n    -webkit-transform: rotate(360deg);\r\n    transform: rotate(360deg);\r\n  }\r\n}\r\n\r\n.gallery ul.gallery-images--ul li.gallery-media-photo.loading::before {\r\n  z-index: 999;\r\n  content: \"\";\r\n  display: block;\r\n  position: absolute;\r\n  left: 0;\r\n  top: 0;\r\n  width: 100%;\r\n  height: 100%;\r\n  background-color: rgba(0,0,0,1);\r\n}\r\n\r\n.mode-light .gallery ul.gallery-images--ul li.gallery-media-photo.loading::before  {\r\n  background-color: rgba(255,255,255,1);\r\n}\r\n\r\n.gallery ul.gallery-images--ul li.gallery-media-photo.loading div.loading-spinner,\r\n.gallery ul.gallery-images--ul li.gallery-media-photo div.picture.loading div.loading-spinner {\r\n  position: absolute;\r\n  z-index: 1001;\r\n  top: 50%;\r\n  left: 50%;\r\n  margin: -1rem 0 0 -1rem; /* visually centered */\r\n  -webkit-transform: translate(-50%, -50%);\r\n  transform: translate(-50%, -50%);\r\n}\r\n\r\n/* gallery animation */\r\n.gallery__animation-appear {\r\n  opacity: 0.01;\r\n  -ms-filter: \"progid:DXImageTransform.Microsoft.Alpha(Opacity=1)\";\r\n  -webkit-transform: translateZ(0);\r\n  -moz-transform: translateZ(0);\r\n  transform: translateZ(0);\r\n}\r\n\r\n.gallery__animation-appear-active {\r\n  opacity: 1;\r\n  -ms-filter: \"progid:DXImageTransform.Microsoft.Alpha(Opacity=100)\";\r\n  -webkit-transition: opacity 150ms ease-out;\r\n  -o-transition: opacity 150ms ease-out;\r\n  -moz-transition: opacity 150ms ease-out;\r\n  transition: opacity 150ms ease-out;\r\n}\r\n\r\n.gallery__image-transition--backwards-enter {\r\n  opacity: 0.01;\r\n  -ms-filter: \"progid:DXImageTransform.Microsoft.Alpha(Opacity=1)\";\r\n  -webkit-transform: translateX(-5px);\r\n  -moz-transform: translateX(-5px);\r\n  -ms-transform: translateX(-5px);\r\n  -o-transform: translateX(-5px);\r\n  transform: translateX(-5px)\r\n}\r\n\r\n.gallery__image-transition--forwards-enter {\r\n  opacity: 0.01;\r\n  -ms-filter: \"progid:DXImageTransform.Microsoft.Alpha(Opacity=1)\";\r\n  -webkit-transform: translateX(5px);\r\n  -moz-transform: translateX(5px);\r\n  -ms-transform: translateX(5px);\r\n  -o-transform: translateX(5px);\r\n  transform: translateX(5px)\r\n}\r\n\r\n.gallery__image-transition--backwards-enter-active,\r\n.gallery__image-transition--forwards-enter-active {\r\n  opacity: 1;\r\n  -ms-filter: \"progid:DXImageTransform.Microsoft.Alpha(Opacity=100)\";\r\n  -webkit-transform: translateX(0);\r\n  -moz-transform: translateX(0);\r\n  -ms-transform: translateX(0);\r\n  -o-transform: translateX(0);\r\n  transform: translateX(0);\r\n  -webkit-transition: opacity 150ms ease-out, -webkit-transform 150ms ease-out;\r\n  transition: opacity 150ms ease-out, -webkit-transform 150ms ease-out;\r\n  -o-transition: opacity 150ms ease-out, -o-transform 150ms ease-out;\r\n  -moz-transition: opacity 150ms ease-out, transform 150ms ease-out, -moz-transform 150ms ease-out;\r\n  transition: opacity 150ms ease-out, transform 150ms ease-out;\r\n  transition: opacity 150ms ease-out, transform 150ms ease-out, -webkit-transform 150ms ease-out, -moz-transform 150ms ease-out, -o-transform 150ms ease-out\r\n}\r\n\r\n.gallery__image-transition--backwards-leave,\r\n.gallery__image-transition--forwards-leave {\r\n  opacity: 1;\r\n  -ms-filter: \"progid:DXImageTransform.Microsoft.Alpha(Opacity=100)\";\r\n  -webkit-transform: translateZ(0);\r\n  -moz-transform: translateZ(0);\r\n  transform: translateZ(0)\r\n}\r\n\r\n.gallery__image-transition--backwards-leave-active,\r\n.gallery__image-transition--forwards-leave-active {\r\n  opacity: 0.01;\r\n  -ms-filter: \"progid:DXImageTransform.Microsoft.Alpha(Opacity=1)\";\r\n  -webkit-transition: opacity 150ms ease-out;\r\n  -o-transition: opacity 150ms ease-out;\r\n  -moz-transition: opacity 150ms ease-out;\r\n  transition: opacity 150ms ease-out\r\n}\r\n\r\n@media (max-width: 743px) {\r\n  .gallery-figcaption--info {\r\n    padding-bottom: 1rem;\r\n    padding-left: 1rem;\r\n    padding-right: 1rem;\r\n  }\r\n\r\n  .gallery-figcaption .gallery-thumbnails--toggle {\r\n    display: none;\r\n  }\r\n}\r\n\r\n@media (min-width: 744px) {\r\n  .gallery .gallery-photo {\r\n    padding-bottom: 67%;\r\n  }\r\n\r\n  .gallery-figcaption--info {\r\n    padding-top: 0.5rem;\r\n    padding-bottom: 0.5rem;\r\n  }\r\n}\r\n\r\n@media (min-width: 1128px) {\r\n  .gallery .gallery-control {\r\n    width: 75px;\r\n    font-size: medium;\r\n  }\r\n}\r\n";
styleInject(css);

var _jsxFileName$b = "O:\\www\\www\\react\\react-bnb-gallery\\src\\ReactBnbGallery.js";
var propTypes$b = forbidExtraProps(_objectSpread({}, galleryPropTypes, {
  leftKeyPressed: PropTypes.func,
  onClose: PropTypes.func,
  rightKeyPressed: PropTypes.func,
  show: PropTypes.bool,
  keyboard: PropTypes.bool,
  opacity: opacityValidation,
  backgroundColor: PropTypes.string,
  zIndex: nonNegativeInteger
}));

var defaultProps$b = _objectSpread({}, galleryDefaultProps, {
  leftKeyPressed: noop,
  onClose: noop,
  rightKeyPressed: noop,
  show: false,
  keyboard: true,
  opacity: DEFAULT_OPACITY,
  backgroundColor: DEFAULT_COLOR,
  zIndex: DEFAULT_Z_INDEX
});

var ReactBnbGallery =
/*#__PURE__*/
function (_Component) {
  _inherits(ReactBnbGallery, _Component);

  function ReactBnbGallery(props) {
    var _this;

    _classCallCheck(this, ReactBnbGallery);

    _this = _possibleConstructorReturn(this, _getPrototypeOf(ReactBnbGallery).call(this, props));
    var photos = _this.props.photos;
    _this.state = {
      photos: null
    };
    _this.gallery = React.createRef();
    _this.close = _this.close.bind(_assertThisInitialized(_assertThisInitialized(_this)));
    _this.onKeyDown = _this.onKeyDown.bind(_assertThisInitialized(_assertThisInitialized(_this)));
    return _this;
  }

  _createClass(ReactBnbGallery, [{
    key: "onKeyDown",
    value: function onKeyDown(event) {
      if (/input|textarea/i.test(event.target.tagName)) {
        return;
      }

      switch (event.which) {
        case ESC_KEYCODE:
          event.preventDefault();
          this.close();
          break;

        case ARROW_LEFT_KEYCODE:
          event.preventDefault();
          this.gallery.current.prev();
          break;

        case ARROW_RIGHT_KEYCODE:
          event.preventDefault();
          this.gallery.current.next();
          break;

        default:
      }
    }
  }, {
    key: "getModalOverlayStyles",
    value: function getModalOverlayStyles() {
      var _this$props = this.props,
          opacity = _this$props.opacity,
          backgroundColor = _this$props.backgroundColor;
      return {
        opacity: opacity,
        backgroundColor: backgroundColor
      };
    }
  }, {
    key: "close",
    value: function close() {
      var onClose = this.props.onClose;
      onClose();
    }
  }, {
    key: "render",
    value: function render() {
      var _this$props2 = this.props,
          show = _this$props2.show,
          phrases = _this$props2.phrases,
          keyboard = _this$props2.keyboard,
          light = _this$props2.light,
          zindex = _this$props2.zIndex;
      var photos = this.state.photos;

      if (!show) {
        return null; // nothing to return
      }

      var galleryProps = omit(this.props, ['onClose', 'leftKeyPressed', 'rightKeyPressed', 'show', 'photos', 'opacity', 'backgroundColor', 'zIndex', 'keyboard']);
      var modalProps = {
        className: classnames(['gallery-modal', light && 'mode-light']),
        role: 'dialog',
        tabIndex: -1,
        onKeyDown: keyboard && this.onKeyDown,
        zindex: zindex
      }; // modal overlay customization styles

      var galleryModalOverlayStyles = this.getModalOverlayStyles();
      return React.createElement(Portal, {
        __source: {
          fileName: _jsxFileName$b,
          lineNumber: 167
        },
        __self: this
      }, React.createElement(FocusTrap, {
        __source: {
          fileName: _jsxFileName$b,
          lineNumber: 168
        },
        __self: this
      }, React.createElement("div", _extends({}, modalProps, {
        __source: {
          fileName: _jsxFileName$b,
          lineNumber: 169
        },
        __self: this
      }), React.createElement("div", {
        style: galleryModalOverlayStyles,
        className: "gallery-modal--overlay",
        __source: {
          fileName: _jsxFileName$b,
          lineNumber: 170
        },
        __self: this
      }), React.createElement("div", {
        className: "gallery-modal--container",
        __source: {
          fileName: _jsxFileName$b,
          lineNumber: 174
        },
        __self: this
      }, React.createElement("div", {
        className: "gallery-modal--table",
        __source: {
          fileName: _jsxFileName$b,
          lineNumber: 175
        },
        __self: this
      }, React.createElement("div", {
        className: "gallery-modal--cell",
        __source: {
          fileName: _jsxFileName$b,
          lineNumber: 176
        },
        __self: this
      }, React.createElement("div", {
        className: "gallery-modal--content",
        __source: {
          fileName: _jsxFileName$b,
          lineNumber: 177
        },
        __self: this
      }, React.createElement("div", {
        className: "gallery-modal--close",
        __source: {
          fileName: _jsxFileName$b,
          lineNumber: 178
        },
        __self: this
      }, React.createElement(GalleryCloseButton, {
        onPress: this.close,
        light: light,
        __source: {
          fileName: _jsxFileName$b,
          lineNumber: 179
        },
        __self: this
      })), React.createElement("div", {
        className: "gallery-content",
        __source: {
          fileName: _jsxFileName$b,
          lineNumber: 184
        },
        __self: this
      }, React.createElement("div", {
        className: "gallery-top",
        __source: {
          fileName: _jsxFileName$b,
          lineNumber: 185
        },
        __self: this
      }, React.createElement("div", {
        className: "gallery-top--inner",
        __source: {
          fileName: _jsxFileName$b,
          lineNumber: 186
        },
        __self: this
      })), React.createElement(Gallery, _extends({
        phrases: phrases,
        ref: this.gallery,
        photos: photos
      }, galleryProps, {
        __source: {
          fileName: _jsxFileName$b,
          lineNumber: 188
        },
        __self: this
      }))))))))));
    }
  }], [{
    key: "getDerivedStateFromProps",
    value: function getDerivedStateFromProps(props, state) {
      if (props.photos !== state.photos) {
        return {
          photos: getPhotos(props.photos)
        };
      }

      return null;
    }
  }]);

  return ReactBnbGallery;
}(Component);

ReactBnbGallery.propTypes = propTypes$b;
ReactBnbGallery.defaultProps = defaultProps$b;

export default ReactBnbGallery;
